### ISSUE TITLE
test_umath.test_nextafter_0 fails on the windows subsystem for linux

### ISSUE THREAD

**Issue by eric-wieser** [numpy/numpy#8726]
Simplest test case:
```
>>> tiny = np.nextafter(np.float128(0), np.float128(1))
3.6451995318824746025e-4951
>>> tiny * 1
0.0
```

**Comment by matthew-brett** [comment]
What system is this on?

**Comment by eric-wieser** [comment]
Intel i5-5200

**Comment by matthew-brett** [comment]
Isn't that a failure of IEEE 754 compliance?

**Comment by eric-wieser** [comment]
Perhaps? Incidentally, building directly on windows doesn't enable `np.float128`, which makes debugging this harder

**Comment by matthew-brett** [comment]
I wonder if it's possible to check whether we see the same behavior in a tiny C program?

**Comment by charris** [comment]
I don't think any of the windows compilers support float128 as a matter of Microsoft policy. I'm surprised you can do that.

**Comment by eric-wieser** [comment]
@charris: I'm not using a windows compiler through, because I'm within a ubuntu VM.

**Comment by charris** [comment]
So the nextafter call goes to a ubuntu library?

**Comment by eric-wieser** [comment]
Nextafter is working correctly I think - the issue is the multiply

**Comment by eric-wieser** [comment]
Seems to work fine in C:

#include <stdio.h>

```
int main() {
    char data[16] = {0};
    data[0] = 1;

    __float128 f = *(__float128*) data;

    printf("%d\n", f == 0);
    printf("%d\n", f*1 == 0);
    return 0;
}
```

Prints `0` twice.

**Comment by matthew-brett** [comment]
Can you replicate the `3.6451995318824746025e-4951 * 1 == 0` multiplication in C?

**Comment by eric-wieser** [comment]
@matthew-brett: I think that's what the above does? I can't use that literal in C though, it gets truncated to 0

**Comment by eric-wieser** [comment]
This works fine, in C++:

```
//  g++ gh8726.cpp -o gh8726 -std=gnu++11 -lquadmath

#include <iostream>
#include <quadmath.h>

using namespace std;

int main() {
	__float128 f = 3.6451995318824746025e-4951q;

	cout << (f == 0) << endl;
	cout << (f*1 == 0) << endl;
}
```

**Comment by matthew-brett** [comment]
Meaning, that these are both 0 ? So the literal is still getting set to 0?

**Comment by matthew-brett** [comment]
OK - so numpy's multiply is at fault?

**Comment by eric-wieser** [comment]
This gets weirder:

```python
>>> tiny64 = np.nextafter(np.float64(0), 1)
>>> tiny128 = np.nextafter(np.float128(tiny64), 0)
>>> tiny128 + 0 == tiny128
False
```

**Comment by eric-wieser** [comment]
Yep, there's some truncation going on

```python
>>> tiny128 == tiny64
False
>>> tiny128 + 0 == tiny64
True
```

**Comment by matthew-brett** [comment]
Is this something to do with the CPU x86 extended precision flag getting flipped I wonder?

**Comment by eric-wieser** [comment]
I think the bug here is the difference between `__float128` and `long double`

**Comment by matthew-brett** [comment]
As far as I know, numpy doesn't use `__float128`, or `quadmath` but only `long double`.   Confusingly, `long double` is named `float128` on Intel platforms, even though it is stored as an extended precision 80-bit float, with packing out to 128 bits.  The value of `tiny` shows that numpy has determined that `long double` is the float80 / float128 type.  But - for some reason - the calculations are done at `double` / `float64` precision.  Chuck - do you know why this is?

**Comment by matthew-brett** [comment]
Try this:

```
//  g++ gh8726.cpp -o gh8726 -std=gnu++11

#include <iostream>
#include <cmath>

using namespace std;

int main() {
	long double f;
	f = std::nextafter(0.0L, 1.0L);

	cout << f << endl;
	cout << (f == 0) << endl;
	cout << (f * 1) << endl;
}
```

I'm guessing you get the same answer as I do on OSX:

```
3.6452e-4951
0
3.6452e-4951
```

**Comment by eric-wieser** [comment]
Yep, output is the same

**Comment by matthew-brett** [comment]
@charris - any insight into what would trigger float64 precision on multiply / add etc of float128 = 80 bit long double?

**Comment by charris** [comment]
I'd like to know what type `tiny` and the result  are and how numpy is installed. I'm also curious as to why a vm is used together with the subsystem for linux. And does the same problem exist for arrays if tiny is a scalar. My first guess is that the python float is taking precedence and converting tiny to float, second is that the result it being converted for some reason. however, those are just unlikely speculations that come to mind without knowing more about the setup.

**Comment by matthew-brett** [comment]
For this script

```
import numpy as np

zeros = np.zeros(2, np.dtype(np.longdouble))
ones = np.ones(2, np.dtype(np.longdouble))
tinies = np.nextafter(zeros, ones)
tinies_t1 = tinies * 1
print(tinies_t1, tinies_t1.dtype)
```

on a not-problematic OSX machine, I get:

```
[ 3.6451995e-4951  3.6451995e-4951] float128
```

My prediction is that @eric-wieser will get this on the Windows subsystem:

```
[ 0.0  0.0] float128
```

consistent with my hypothesis that all basic operations are being done at float64 precision, where both the inputs and outputs are float128 (== padded float80).

**Comment by matthew-brett** [comment]
I'm guessing that this is something to do with the [FPU control world](http://www.efg2.com/Lab/Library/Delphi/MathFunctions/FPUControlWord.Txt) setting on the x86 chip; specifically, that FPU precision control field is set to use double (float64) precision rather than extended (80-bit) precision.  @cournape - I seem to remember that you know this world ... ?

**Comment by eric-wieser** [comment]
> My prediction is that @eric-wieser will get this on the Windows subsystem:

Can confirm

> I'd like to know what type tiny and the result are and how numpy is installed.

Numpy was installed from source with `./runtests.py -i`

> I'm also curious as to why a vm is used together with the subsystem for linux.

I mispoke, I'm simply running within the linux subsystem, which I likely-incorrectly referred to as a VM.

> And does the same problem exist for arrays if tiny is a scalar. My first guess is that the python float is taking precedence and converting tiny to float, second is that the result it being converted for some reason.

I seem to get the same result whether the operands are scalars, arrays, `float`s, or `np.float128`s

**Comment by njsmith** [comment]
Yeah, I'm not sure what we can do here -- if it's broken on WSL but works on real!Linux then by definition it's a bug in WSL. And if 80-bit float support is broken then there's not much we can do to work around that – we're not going to add soft-float emulation or something. Even if it's "just" that they have the fp control flag set wrong then I don't see how we can reasonably compensate -- in particular, it's a per-thread setting, so even if we try to flip it back to the correct state then we'll show even more obscure bugs in threaded code, where new threads revert to the broken behavior. Or else we'd have to check and reset the flag before every operation, which would require a lot of error-prone work and probably slow down real!Linux too much anyway.

Honestly I'm not sure why we support long double at all; they're confusing and flaky and a huge source of maintenance hassles, and not really that useful in the first place.

**Comment by njsmith** [comment]
BTW it should be pretty straightforward to confirm whether WSL is setting up the FPU correctly – just print the FPU control word on WSL and real!Linux, and if the relevant bits are different then WSL is buggy. (This is a very plausible bug, too, since it's a place where Windows and Linux definitely do differ in their defaults, and where the effects would mostly fly under the radar.)

I don't have access to WSL, but something vaguely like this should work (untested):
```c
#include <stdio.h>
#include <fpu_control.h>

int main(int argc, char** argv) {
    fpu_control_t cw;
    _FPU_GETCW(cw);
    printf("%x\n", cw);
}
```

It looks like this is the place to file bugs on WSL: https://github.com/Microsoft/BashOnWindows/issues

**Comment by eric-wieser** [comment]
@njsmith: I don't think that getting FPU flags at startup will help - remember that this bug doesn't appear in a minimal C or C++ program, so at the very least numpy is switching on a define missed by WSL, or configuring the FPU in a way that breaks WSL.

**Comment by njsmith** [comment]
Probably worth double-checking anyway. How confident are you that your minimal C++ program didn't optimize `f * 1` into `f`?

Otherwise I think the only thing numpy does with the FPU is make calls like:
```c
#include <fenv.h>

    int fpstatus = fetestexcept(FE_DIVBYZERO | FE_OVERFLOW |
                                FE_UNDERFLOW | FE_INVALID);

    feclearexcept(FE_DIVBYZERO | FE_OVERFLOW |
                  FE_UNDERFLOW | FE_INVALID);

    /* and similarly for the other flags above */
    feraiseexcept(FE_DIVBYZERO);
```
I guess you could play around with those to see if any of them break stuff on WSL. (see `numpy/core/src/npymath/ieee754.c.src` for details)

**Comment by eric-wieser** [comment]
> How confident are you that your minimal C++ program didn't optimize f * 1 into f?

Is `gcc` allowed to do that with `-O0`?

**Comment by njsmith** [comment]
`gcc` is allowed to interpret its command line arguments however it wants :-)

From `fpu_control.h`:
```
 *     15-13    12  11-10  9-8     7-6     5    4    3    2    1    0
 * | reserved | IC | RC  | PC | reserved | PM | UM | OM | ZM | DM | IM
[...]
 * PC: Precision control
 * 11 - round to extended precision
 * 10 - round to double precision
 * 00 - round to single precision
```

So it looks like precision control is bits 8-9, and `bin((0x27f >> 8) & 0x3)` gives me `10`, i.e. you seem to have extended precision disabled. Running my test program on real Linux I get 37f, which corresponds to `PC = 11`, extended precision enabled.

**Comment by njsmith** [comment]
Filed upstream: https://github.com/Microsoft/BashOnWindows/issues/1748

**Comment by matthew-brett** [comment]
@eric-wieser - I think Nathaniel is right - the `f * 1` is getting optimized away - try:

```
/* gcc -lm check2.c */

#include <stdio.h>
#include <math.h>

int main (void) {
    long double f, f2;

    f = nextafterl(0.0L, 1.0L);
    f2 = f * 2;

    printf("%Le\n", f);
    printf("f == 0? %d\n", f == 0);
    printf("%Le\n", f2);
}
```

I think you'll get

```
3.645200e-4951
f == 0? 0
0.000000e+00
```
